TFNetworkLayer.py

```diff
@@ -904,8 +908,12 @@ def batch_norm(self, data,
       https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/batch_norm.py
     """
     with tf.variable_scope("batch_norm"):
-      x = data.get_placeholder_flattened(keep_dims=True)  # shape (time',...)
-      mean, variance = tf.nn.moments(x, axes=[0], keep_dims=True)
+      if masked_time:
+        x = data.get_placeholder_flattened(keep_dims=True)
+        mean, variance = tf.nn.moments(x, axes=[0], keep_dims=True)
+      else:
+        x = data.placeholder
+        mean, variance = tf.nn.moments(x, axes=data.get_axes(exclude_feature=True), keep_dims=True)
```

Comment:
This is not consistent, and thus wrong.
E.g. consider you have `data` with (batch) shape (B, T, H, D).
In the `masked_time` case, you calculate mean/var over (B, T), i.e. the mean/var shape is (1, 1, H, D).
In the other case, you calculate mean/var over (B,T,H), i.e. the mean/var shape is (1,1,1,D).

Reply:
The question is also, what do you really want to have here? If you need that new behavior (mean/var over (B,T,H)), then that should be yet another option.
Btw, in both cases (`masked_time` or not), you can use the same `axes` argument, because it should be the same.

Reply:
Also, I wonder, why did you not catch this error in your test case? That means that your test case is likely broken.

Reply:
> In the `masked_time` case, you calculate mean/var over (B, T), i.e. the mean/var shape is (1, 1, H, D).
No, the mean/var shape is (1, 1, 1, D), which is a default behavior. `data.get_placeholder_flattened(keep_dims=True)` returns (BxTxH, D) tensor.

Reply:
Oh yes, you are right.